### DISCUSSION TITLE
"Annual" in addition to (instead of?) 12-month renewal period

### DISCUSSION THREAD

**Discussion by TiCampbell** [Get-Your/Get-Your#304] [General]
@penoptech yesterday, I was introduced to the concept that these programs are technically available annually; the 12-month renewal period allows for rolling renewals to not inundate the program coordinator all at once, but applicants are allowed to renew their benefits each calendar year. We've already received some questions from applicants wondering when they can reapply and receive their 2024 Get FoCo benefits.

I have two questions for you on this:

1. Would it be possible to have a calendar-year annual renewal *in addition to* the upcoming `last_completed_at` 12-month renewal period? So if somebody applied in October 2023, the 12-month renewal would calculate a renewal requirement of October 2024, but each program would be *available* for renewal on January 1, 2024. This would allow administration to trigger a user renewal as soon as new-year benefits become available, but also would enable creation of a 'Renew Now' button (#303) for user self-service.
2. Is there a straightforward way to genericize the two renewal types? It would be great to have this be program-specific (in the `IQProgramRD` table), so we could specify that a program has either calendar-year or number-of-months renewal while another program has only number-of-months renewal (and would it be possible to have a calendar-year renewal *only*? I would think we would disallow this or else there would be no reminder workflow).

**Comment by TiCampbell** [comment]
~~Also, a calendar-year renewal would likely need something like a `next_renewal_available_at` field in the `IQProgram` table that is set to January 1 at midnight for the next year.~~

See reply.

**Reply by TiCampbell** [reply]
After talking with Nina, the easier way to handle calendar-year renewals is to just only allow user renewal outside of the 12-month rolling during the *following calendar year*. For a 'Renew Now' button, this would just only show the button if `now().year = last_completed_at.year + 1`

**Comment by TiCampbell** [comment]
We received new information from Recreation that they actually use a calendar-year renewal (instead of the rolling 12-month that we implemented), so we'll need to have a calendar-year implementation (and this necessitates the 'renew now' button referenced in the description).

**Comment by TiCampbell** [comment]
Regarding (2) in the description:

I think a larger question with all this 'calendar-year-renewal' vs 'annual renewal' talk and having both options, is do we need to change the way we're allowing renewal intervals to be defined for each program? It seems like we could still keep the option for >1-year renewals, we would just need to force renewal intervals to be multiples of a year so that we could use `now().year = last_completed_at.year + <renewal_interval_year>` to allow 'calendar-year renewals' (e.g. the Renew Now button).

Also, should 'calendar-year renewal' be an option to disable, i.e. is 'calendar-year renewal' a Boolean that enables the Renew Now button or should the Renew Now option always be available at the beginning of each year (or however long, based on the minimum renewal interval). I think having a Boolean would be helpful for flexibility.

I propose that we change `renewal_interval_month` to `renewal_interval_year` and keep it as a null-allowed integer, so that admins are required to define specific years for each renewal interval (when they exist), and that the option for users to renew from the beginning of the year of their renewal date be added as a Boolean 'enable calendar-year renewal' to the **universal site settings** (not program-specific) table (#122).

**Comment by penoptech** [comment]
@TiCampbell - ~Do we still want to send renewal notifications to users that are enrolled in programs with calendar-year renewals? If so is it safe to assume all programs that have calendar-year renewals start on January 1st?~

For your `calendar-year-renewal` boolean... Are you suggesting we add a boolean column to the `IQProgramsRD` table and then add logic to the `check_if_user_needs_to_renew` function to see if any of the user's programs need to be renewed on January 1st? 

Doing that will coerce a user into a rolling renewal starting on January 1st if they have a program with that boolean checked. That approach makes a lot of sense to me. We can keep everything that supports the rolling renewal functionality with only a few minor modifications to support calendar renewals.

You mentioned:

>  ...we would just need to force renewal intervals to be multiples of a year so that we could use now().year = last_completed_at.year + <renewal_interval_year> to allow 'calendar-year renewals' (e.g. the Renew Now button).

By managing if an IQ program has a rolling or calendar renewal, we might not need the "Renew Now" button. If any of the user's programs have `calendar_renewal = True` and the `app_iqprogram.applied_at.year < now().year`, we can return `True` from the `check_if_user_needs_to_renew` function and display the existing renew button on the dashboard.

**Reply by TiCampbell** [reply]
I do like the administration simplicity of coercing users to a calendar-year renewal, but I was thinking a 'Renew Now' button to allow flexibility for the users. If we do a `calendar_renewal` Boolean, I could see cases where someone completes their initial application in December, but then the system would force them to gather their paperwork again less than a month later. A 'Renew Now' button would give them the option to go through the renew, but wouldn't force them to.

There are two other factors that I haven't been able to enunciate on why I like the self-serviceness of the 'Renew Now':
- I've been having to do a lot of manual updates and resets because despite the team's best efforts, there are still outliers to our renewals plan. Instructing the program coordinators to guide users to a 'Renew Now' button when these outliers arise instead of elevating them to the admins sounds fantastic.
- Recreation actually has some leeway in their calendar-year renewal, that I don't want to try to account for. The way it's set up is that if a user applies for the reduced-rate pass from January through October, they have until the end of that calendar year to buy the pass; if they apply in November or December, they have until the end of the *next* year (so as not to force people to immediately buy a pass).

But if we go this route, I think it's too complicated to have a `calendar_renewal` like I had been thinking. Nina's idea of just hiding the 'Renew Now' button until January 1 of the following year is a cleaner way of doing it. The button itself is still a different mechanism (it wouldn't be able to use the same `check_if_user_needs_to_renew`) but the renewal flow would be the same once the user clicked the button.

**Reply by TiCampbell** [reply]
@penoptech When talking to Nina yesterday I learned about another aspect I hadn't considered, and I think it's now a *need* to have the Renew Now option instead of the coercing into a 'renewal required' state. Apparently the administrative work is so high for some of these benefits that one of our programs is very backed up just with the new applicants this year, and we're starting to get calls from residents wondering what's taking so long.

If we start coercing users into an January 1 renewal, we're going to overwhelm the program coordinators, and also (what I didn't think about) degrade the applicant experience by dragging out their enrollment process.

What do you think?

**Reply by penoptech** [reply]
@TiCampbell My thought is that if it's a program that allows for renewals starting at the beginning of every calendar year that the flood of applicants might happen regardless.

You bring up a good point though. My suggested approach could still be an issue, because program renewals for a "smaller" (lower resourced) program could be overwhelmed when coerced into a "larger" (higher resourced) program's renewal schedule.

**Reply by penoptech** [reply]
After talking with @TiCampbell, we're both in agreement that the cons of coercing programs into the same renewal schedule out way the benefits of the approach. 

For the Renew Now button, the button will be available the whole time on the dashboard and if a user happens to log in they can go through the application renewal.

To an earlier point of Tim's, we need to warn users when they use the 'Renew Now' function that it can adversely impact their benefits depending on what information changes for them.
